# Исследование надежности заемщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание проекта.**

По представленным статичтическим данным о платежеспособности клиентов банка необходимо провести исследование- влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования — проверить гипотезы:**

Существует ли зависимость:

- между наличием детей и возвратом кредита в срок;
- между семейным положением и возвратом кредита в срок;
- между уровнем дохода и возвратом кредита в срок;
- между целью кредита и его возврат в срок.

**Ход исследования**

Данные о поведении пользователей (статистика о платёжеспособности клиентов)- в файле /datasets/data.csv. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных.

Проверим данные на ошибки и оценим их влияние на исследование. Затем, на этапе предобработки поищем возможность исправить самые критичные ошибки данных.

Таким образом, исследование пройдёт в три этапа:

- Обзор данных.
- Предобработка данных.
- Проверка гипотез.

Это первая часть проекта, она будет проверена автоматически. Вторую часть проверит ревьюер. 

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
data=pd.read_csv('/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [56]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for d in data['income_type'].unique():
    data.loc[(data['income_type'] == d) & (data['total_income'].isna()),'total_income']=\
    data.loc[(data['income_type'] == d),'total_income'].median()
  

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].median()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data=data[(data['children'] != -1)&(data['children'] != 20)]# ваш код здесь

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждому типу занятости `income_type`.**

In [11]:
for d in data['income_type'].unique():
    data.loc[(data['income_type'] == d) & (data['days_employed'].isna()),'days_employed']=\
    data.loc[(data['income_type'] == d),'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype('int')

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()


**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их. Сбрасывать индексы после удаления строк дубликатов с помощью `reset_index(drop=True)` здесь не требуется.**

In [15]:
data['education'].duplicated().sum()  

21397

In [16]:
data['education'].drop_duplicates()

0                    высшее
1                   среднее
13      неоконченное высшее
31                начальное
2963         ученая степень
Name: education, dtype: object

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
# создадим функцию categorize_income()
def categorize_income(total_income):
    
    if 0<= total_income <= 30000:
        return 'E' 
    elif 30001<= total_income <= 50000:
        return 'D'  
    elif 50001<= total_income <= 200000:
        return 'C'  
    elif 200001<= total_income <= 1000000:
        return 'B'    
    elif total_income >= 1000001:
        return 'A'
        

In [18]:
# применим функцию методом apply()
data['total_income_category']= data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:

def categorize_purpose(purpose):
   
    if 'образ' in purpose:
        return 'получение образования'
    if 'свадь' in purpose:
        return 'проведение свадьбы'
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'недвиж' or 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свадь' in purpose:
        return 'проведение свадьбы'

In [21]:

data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data['purpose_category']. unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [38]:
def children_in_category(row):
    row=row['children']
    if row == 1:
        return 'один ребенок'
    if row == 2:
        return 'два ребенка'
    else:
        return 'многодетные заемщики'
data['children_category'] = data.apply(children_in_category, axis=1)    
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category     children_category  
0      операции с недвижимостью          один ребенок  
1        операции с автомобилем          один ребенок  
2      операции с недвижимостью  многодетные заемщики  
3         получение образования  многодетные заемщики  
4            проведение свадьбы  многодетные заемщики  
...                         ...                   ...  
21520  операции с недвижимостью          один ребенок  
21521    операции с автомобилем  многодетные заемщики  
21522  операции с недвижимостью          один ребенок  
21523    операции с автомобилем  многодетные заемщики  
21524    операции с автомобилем           два ребенка  

[21402 rows x 15 columns]

In [24]:
Children_dependence = data.groupby('children_category')['debt'].agg(['count','sum','mean'])

In [25]:
Children_dependence.rename(columns={'count': 'Всего заемщиков', 'sum': 'Количество должников','mean': 'Доля должников'},inplace=True)

In [26]:
Children_dependence

Всего заемщиков  Количество должников  Доля должников
children_category                                                          
два ребенка                      2055                   194        0.094404
многодетные заемщики            14529                  1094        0.075298
один ребенок                     4818                   444        0.092154

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [27]:
Family_st_dependence = data.groupby('family_status')['debt'].agg(['count','sum','mean'])

In [28]:
Family_st_dependence.rename(columns={'count': 'Всего заемщиков', 'sum': 'Количество должников','mean': 'Доля должников'},inplace=True)

In [81]:
Family_st_dependence

Всего заемщиков  Количество должников  Доля должников
family_status                                                               
Не женат / не замужем             2799                   273        0.097535
в разводе                         1189                    84        0.070648
вдовец / вдова                     952                    63        0.066176
гражданский брак                  4160                   385        0.092548
женат / замужем                  12302                   927        0.075354

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [35]:
Total_income_dependence = data.groupby('total_income_category')['debt'].agg(['count','sum','mean'])

In [36]:
Total_income_dependence.rename(columns={'count': 'Всего заемщиков', 'sum': 'Количество должников','mean': 'Доля должников'},inplace=True)

In [37]:
Total_income_dependence

Всего заемщиков  Количество должников  Доля должников
total_income_category                                                       
A                                   25                     2        0.080000
B                                 5014                   354        0.070602
C                                15992                  1353        0.084605
D                                  349                    21        0.060172
E                                   22                     2        0.090909

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

In [32]:
Purpose_dependence = data.groupby('purpose_category')['debt'].agg(['count','sum','mean'])

In [33]:
Purpose_dependence.rename(columns={'count': 'Всего заемщиков', 'sum': 'Количество должников','mean': 'Доля должников'},inplace=True)

In [34]:
Purpose_dependence

Всего заемщиков  Количество должников  \
purpose_category                                                  
операции с автомобилем               4288                   400   
операции с недвижимостью            10780                   780   
получение образования                3997                   369   
проведение свадьбы                   2337                   183   

                          Доля должников  
purpose_category                          
операции с автомобилем          0.093284  
операции с недвижимостью        0.072356  
получение образования           0.092319  
проведение свадьбы              0.078306

**Вывод:**

По результатам исследования можно сделать следующие выводы:

1.Наблюдается зависимость между количеством детей и возвратом кредита в срок.

Люди, имеющие 1,2 ребенка имеют самую большую долю по просрочке 0.092-0.094! Я думаю это связано с тем, что они часто прибегают к различным кредитам(образование детей,потребительские кредиты).

Доля людей не имеющих детей составляет 0.075, думаю это связано с другой финансовой нагрузкой.

У людей имеющих более 3х детей, доля просрочек меньше 0.075,скорее всего это связано с тем, что эта категория рассматривается банком как нежелательные заемщики и им отказывают в выдаче кредитов.

2.Зависимость между семейным положением и возвратом кредита в срок.

Самая большая доля просрочки у людей не женатых/ не замужем(0.098) и живущих в гражданском браке (0.093). Я думаю это связано с тем, что это относительно молодая, по возрасту группа людей, который часто берут кредиты, но еще не умеют граммотно планировать свои доходы и расходы.

Доля должников среди людей в разводе (0.071) и вдова/вдовец (0.066). Думаю это люди, который умеют хорошо планировать свои доходы и расходы и подходят к взятию кредитов более ответственно.

3.Зависимость между уровнем дохода и возвратом кредита в срок.

Исследование показала, что люди с высокими доходами не имеют просрочек, думаю это связано с тем, что у них нет причин для взятия кредитов. Доля людей с низкими доходами тоже не высока, из-за того что банки могут отказать им в выдаче кредитов.

4.Как разные цели кредита влияют на его возврат в срок.

Самая больщая доля просрочек по кредитам приходится на категории образования (0.095) и автомобили (0.092). Скорее всего это связано с тем, что это самые частые причины по которым люди берут кредиты и банки предлагают по ним выгодные условия.
Небольшую долю просрочек состовляют кредиты, связанные с покупкой жилья (0.055),причина может быть в том, что банки предъявляют к этой категории больше требований и не всем одобряют такие кредиты. 
